// foxie-themes/controller/overide/themeManifest.js
var foxieData = require("foxie-data")

function closeMatch(iterate, ext) {
	var fileExt = "."+iterate.split(".").pop();
	if (fileExt == ext) return true;
	else return false;
}

function $class(themeName) {
	this.make = function (res, themePath) {

		this.tmpArr = []
		this.themeName = themePath.split("/").pop()
		this.tmpArr[this.themeName] = {}
		var styles = this.tmpArr[this.themeName]["styles"] = {}
		var scripts = this.tmpArr[this.themeName]["scripts"] = {}
		var data = this.tmpArr[this.themeName]["data"] = {}
		var plain = this.tmpArr[this.themeName]["plain"] = {}
		var templates = this.tmpArr[this.themeName]["templates"] = {}
		var images = this.tmpArr[this.themeName]["images"] = {}

		styles["scss"] = []
		styles["less"] = []
		styles["styl"] = []
		styles["css"] = []
		data["json"] = []
		data["yml"] = []
		data["xml"] = []
		scripts["js"] = []
		plain["txt"] = []
		plain["md"] = []
		templates["jade"] = []
		templates["handlebars"] = []
		templates["ejs"] = []
		templates["html"] = []
		images["jpeg"] = []
		images["png"] = []
		images["gif"] = []
		images["ico"] = []

		for(i in res){//sort into groups
			if(closeMatch(res[i],".scss"))	styles["scss"].push(res[i])
			if(closeMatch(res[i],".less"))	styles["less"].push(res[i])
			if(closeMatch(res[i],".styl"))	styles["styl"].push(res[i])
			if(closeMatch(res[i],".css"))		styles["css"].push(res[i])
			if(closeMatch(res[i],".json"))	data["json"].push(res[i])
			if(closeMatch(res[i],".yml"))		data["yml"].push(res[i])
			if(closeMatch(res[i],".xml"))		data["xml"].push(res[i])
			if(closeMatch(res[i],".js"))		scripts["js"].push(res[i])
			if(closeMatch(res[i],".txt"))		plain["txt"].push(res[i])
			if(closeMatch(res[i],".md"))		plain["md"].push(res[i])
			if(closeMatch(res[i],".jade"))	templates["jade"].push(res[i])
			if(closeMatch(res[i],".ejs"))		templates["ejs"].push(res[i])
			if(closeMatch(res[i],".hbs"))		templates["handlebars"].push(res[i])
			if(closeMatch(res[i],".html"))	templates["html"].push(res[i])
			if(closeMatch(res[i],".htm"))		templates["html"].push(res[i])
			if(closeMatch(res[i],".jpeg"))	images["jpeg"].push(res[i])
			if(closeMatch(res[i],".jpg"))		images["jpeg"].push(res[i])
			if(closeMatch(res[i],".png"))		images["png"].push(res[i])
			if(closeMatch(res[i],".gif"))		images["gif"].push(res[i])
			if(closeMatch(res[i],".ico"))		images["ico"].push(res[i])
		}

		var handlerPath = "../handlers";

		//do specific things for each file type
		require(handlerPath+"/templates.js")(templates,foxieData.temp,themeName)
		require(handlerPath+"/images.js")(images)
		require(handlerPath+"/plain.js")(plain)
		require(handlerPath+"/scripts.js")(scripts)
		require(handlerPath+"/styles.js")(styles,themeName)
		require(handlerPath+"/data.js")(data)

	}
}

module.exports = function(res,themePath,themeName){
var manifest = new $class(themeName)
	manifest.make(res,themePath)
}
